[docker] improve corecheck integration test (exit codes, fixes)

JulienBalestra opened the pull request:
### What does this PR do?

This PR arrange the organisation of the integration-tests in the docker corechecks.
Add a test support for the check name `docker.exit`.

### Motivation

Improve the coverage of the integration-tests

JulienBalestra committed:
Rename base_test to basemetrics, add a testing suite for the exit codes

xvello commented on test/integration/corechecks/docker/catalog.go:
```diff
@@ -6,13 +6,15 @@
-	composeFiles: make(map[string]string),
+var defaultCatalog catalog
+
+func (c *catalog) append(filename string) {
```
can you rename it `addCompose` as the catalog might hold more than that later?

xvello commented on test/integration/corechecks/docker/exitcode_test.go:
```diff
@@ -0,0 +1,36 @@
+	registerComposeFile("exitcode.compose")
+}
+
+type exitCodeAssertSucceed struct {
```
would be great to document why we use this struct, as other teams might dig into the test if it were to fail

xvello commented on test/integration/corechecks/docker/exitcode_test.go:
```diff
@@ -0,0 +1,36 @@
+
+func TestContainerExit(t *testing.T) {
+	if exitCodeAssertsState.exit0 == false {
+		exitCodeAssertsState.exit0 = sender.AssertServiceCheck(t, "docker.exit", metrics.ServiceCheckOK, "", []string{instanceTag}, "Container exitcode_exit0_1 exited with 0")
```
I guess we could test for more tags, depending on the container's specs. Which ones are to be extracted?

xvello commented on test/integration/corechecks/docker/exitcode_test.go:
```diff
@@ -0,0 +1,36 @@
+		exitCodeAssertsState.exit0 = sender.AssertServiceCheck(t, "docker.exit", metrics.ServiceCheckOK, "", []string{instanceTag}, "Container exitcode_exit0_1 exited with 0")
+	}
+	if !exitCodeAssertsState.exit1 == false {
+		exitCodeAssertsState.exit1 = sender.AssertServiceCheck(t, "docker.exit", metrics.ServiceCheckOK, "", []string{instanceTag}, "Container exitcode_exit1_1 exited with 1")
```
exit codes != 0 should send `metrics.ServiceCheckCritical`. Either the test passes as a false positive or you found a bug in the check. check what `agent check docker` outputs

xvello commented on test/integration/corechecks/docker/main_test.go:
```diff
@@ -26,10 +26,14 @@ var retryDelay = flag.Int("retry-delay", 1, "time to wait between retries (defau
 tags:
-  - integration:test
+  - instanceTag:MustBeHere
+collect_exit_codes: true
```
Please keep the order of docker's `conf.yaml.example` for readability

xvello commented on test/integration/corechecks/docker/basemetrics_test.go:
```diff
@@ -12,29 +12,34 @@ import (
 )
 
+func init() {
+	registerComposeFile("basemetrics.compose")
```
👍

xvello commented on test/integration/corechecks/docker/testdata/exitcode.compose:
```diff
@@ -0,0 +1,14 @@
+version: '2'
+services:
+  exit0:
+    image: "redis:latest"
```
you could use a more lightweight `busybox` image

xvello commented on test/integration/corechecks/docker/testdata/exitcode.compose:
```diff
@@ -0,0 +1,14 @@
+version: '2'
+services:
+  exit0:
```
you should add labels and envs to make sure they are collected as tags

JulienBalestra committed:
[WIP] Use anchor for yaml, deeper coverage for tags

JulienBalestra committed:
[mocksender] Remove the AssertContains and AssertNotContains by a double loop. Improve the testing of the mocksender.
[integration-tests] Fix the Integration tests of exit code ServiceCheck.

xvello commented on pkg/aggregator/mocksender/asserts.go:
```diff
@@ -43,29 +48,33 @@ func (m *MockSender) AssertMetricNotTaggedWith(t *testing.T, method string, metr
 
+// Return a bool value if all the elt of expected are inside the actual array
+func expectedInActual(actual, expected []string) bool {
+	expectedCount := 0
```
You can avoid creating that variable thanks to [loop labels](http://relistan.com/continue-statement-with-labels-in-go/):
```
if e == a {
  continue ITER_EXPECTED
}
return false
```

mfpierre replied:
@xvello I'm not sure it's a good idea to use labels, maybe you'll avoid creating a variable but IMHO you'll lose much more in code readability. (it's not indented + not many people use this syntax, and opens the `goto` door 😬)

JulienBalestra replied:
The implementation looks like that:
```ITER:
	for _, e := range expected {
		for _, a := range actual {
			if e == a {
				continue ITER
			}
		}
		return false
	}
	return true
```
I prefer creating a variable it's more explicit and a common way to implement this kind of search loop

xvello replied:
okay then ;)

xvello commented on pkg/aggregator/mocksender/asserts.go:
```diff
@@ -43,29 +48,33 @@ func (m *MockSender) AssertMetricNotTaggedWith(t *testing.T, method string, metr
 }
 
+// Return a bool value if all the elt of expected are inside the actual array
+func expectedInActual(actual, expected []string) bool {
```
When reading the unit tests, I was surprised at why they'd work as the signature I was infering was `expectedInActual(expected, actual []string)`. Would it make sense to you to swap the arguments?

JulienBalestra replied:
I based the signature of this function on the original one we replaced:
`assert.Contains(t, actual, tag)` but I can swap the arguments of course.

xvello replied:
I'm expecting the first argument to `Contains` to be the one containing the second argument. I'm ok with this name too.

JulienBalestra replied:
I'll change it because it's more obvious in other languages too: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertIn

xvello commented on pkg/aggregator/mocksender/mock_test.go:
```diff
@@ -0,0 +1,49 @@
+	sender.ServiceCheck("docker.exit", metrics.ServiceCheckOK, "", tags, message)
+	sender.AssertCalled(t, "ServiceCheck", "docker.exit", metrics.ServiceCheckOK, "", tags, message)
+	sender.AssertNotCalled(t, "ServiceCheck", "docker.exit", metrics.ServiceCheckOK, "", append(tags, "1"), message)
+	sender.AssertServiceCheck(t, "docker.exit", metrics.ServiceCheckOK, "", tags, message)
```
Can you also add a test assertion for []string{} and tags[:1], to make sure expecting only a subset of tags does work?

xvello commented on test/integration/corechecks/docker/exitcode_test.go:
```diff
@@ -0,0 +1,44 @@
+		"highcardenvtag:exithighenv",
+		"lowcardenvtag:exitlowenv",
+	}
+	sender.AssertServiceCheck(t, "docker.exit", metrics.ServiceCheckOK, "", expectedTags, "Container exitcode_exit0_1 exited with 0")
```
please append the container_name tag to all three asserts

xvello commented on test/integration/corechecks/docker/exitcode_test.go:
```diff
@@ -0,0 +1,44 @@
+	}
+	sender.AssertServiceCheck(t, "docker.exit", metrics.ServiceCheckOK, "", expectedTags, "Container exitcode_exit0_1 exited with 0")
+	for _, check := range []metrics.ServiceCheckStatus{metrics.ServiceCheckWarning, metrics.ServiceCheckCritical, metrics.ServiceCheckUnknown} {
+		sender.AssertServiceCheckNotCalled(t, "docker.exit", check, "", expectedTags, "Container exitcode_exit0_1 exited with 0")
```
Good to test that. To cover a broader scope of calls AssertNotCalled, I'd only assert:
  - "docker.exit"
  - check  (to rename `status` for readability btw)
  - AnythingOfType("string")
  - just []string{"container_name:xxxxx"}
  - AnythingOfType("string")


what do you think?

JulienBalestra committed:
[integration-tests] Cover the ServiceCheck docker.service_up

xvello commented on test/integration/corechecks/docker/testdata/exitcode.compose:
```diff
@@ -0,0 +1,36 @@
+    entrypoint: /bin/sh
+    command: "-c 'exit 1'"
+    labels:
+        low.card.label: *lowcardlabel
```
You could go further with deduplication by directly copying the exit0 as template and overriding just the `command`, see https://github.com/docker/compose/issues/1988#issuecomment-153377186 for an example

xvello reviewed (changes requested):
Good overall improvement, and CI is green 🌮 
Mixed bag of improved coverage and code style suggestions

JulienBalestra committed:
[mocksender] Swap the arguments for expectedInActual, create a new method to assert only one ServiceCheck was called
[integration-tests] Fix the tearDown to remove containers, refactor the compose file

JulienBalestra committed:
[mocksender] Use an AssertServiceCheck in an exclusive way, remove useless arguments in other func tools
[integration-tests] DRY the compose catalog

xvello commented on pkg/aggregator/mocksender/asserts.go:
```diff
@@ -14,64 +14,78 @@ import (
+	return m.Mock.AssertCalled(t, method, metric, mock.AnythingOfType("float64"), mock.AnythingOfType("string"), AssertTagsNotContains(tags))
+}
+
+// Return a bool value if all the elt of expected are inside the actual array
```
s/elt/elements/

xvello replied:
I was ready to ask to move it to the public interface, but in fact users should use AssertTagsContains / AssertTagsNotContains

Could you please move it at the very bottom? I like to keep methods sorted from most useful public to least useful public to private, when possible

xvello commented on pkg/aggregator/mocksender/asserts.go:
```diff
@@ -14,64 +14,78 @@ import (
+	return len(expected) == expectedCount
+}
+
+// AnythingBut match everything except the argument
```
Please document it builds a mock.argumentMatcher

xvello commented on pkg/aggregator/mocksender/asserts.go:
```diff
@@ -14,64 +14,78 @@ import (
 // AssertTagsContains is a mock.argumentMatcher builder to be used in asserts.
 // It allows to check if tags are emitted, ignoring unexpected ones and order.
-func AssertTagsContains(t *testing.T, expected []string) interface{} {
+func AssertTagsContains(expected []string) interface{} {
```
We should find a better name than assert*, maybe MatchTagsContains?
same for all the others

xvello commented on pkg/aggregator/mocksender/mock_test.go:
```diff
@@ -0,0 +1,51 @@
+	assert.True(t, expectedInActual([]string{}, []string{"one", "two"}))
+	assert.True(t, expectedInActual([]string{"one"}, []string{"one", "two"}))
+	assert.True(t, expectedInActual([]string{"one", "two"}, []string{"one", "two"}))
+	assert.True(t, expectedInActual([]string{"one", "two"}[0:0], []string{"one", "two"}))
```
the  `[ ]` construct is good if your array is already allocated somewhere. either directly allocate what you want or pre-allocate []string{"one", "two", "three"} in a var an reuse it

xvello reviewed (changes requested):
Logic is great, doc improvements

JulienBalestra committed:
[mocksender] Cover by unittests the logic of AnythingBut, move private func to the bottom of the file + other style-fix

JulienBalestra commented:
Rebasing on master

xvello reviewed (approved):
![](https://ljdchost.com/6jMbhKn.gif)
